crates/nu-protocol/src/engine/pwd_per_drive.rs

```diff
@@ -0,0 +1,482 @@
+///
+/// Usually if a command opens one file or directory, it uses
+/// nu_path::expand_path_with::<P, Q>(p, r, t) to expand '~','.' etc.; replacing it with
+/// nu_protocol::engine::fs_client::expand_path_with(stack, engine_state, p, r t) will
+/// first check if the path is relative for a drive;
+/// Commands that accept multiple files/directories as parameters usually depend on Glob,
+/// after near future revised Glob collection implementation done, all file system commands
+/// will support PWD-per-drive.
+pub fn expand_path_with<P, Q>(
+    _stack: &Stack,
+    _engine_state: &EngineState,
+    path: P,
+    relative_to: Q,
+    expand_tilde: bool,
+) -> PathBuf
+where
+    P: AsRef<Path>,
+    Q: AsRef<Path>,
+{
+    #[cfg(windows)]
+    if let Some(abs_path) = windows::expand_pwd(_stack, _engine_state, path.as_ref()) {
+        return abs_path;
+    }
+
+    nu_path::expand_path_with::<P, Q>(path, relative_to, expand_tilde)
+}
+
+#[cfg(windows)]
+pub mod windows {
+    use super::*;
+
+    pub trait EnvMaintainer {
+        fn maintain(&mut self, key: String, value: Value);
+        fn provide(&mut self, key: String) -> Option<Value>;
+    }
+
+    impl EnvMaintainer for EngineState {
+        fn maintain(&mut self, key: String, value: Value) {
+            self.add_env_var(key, value);
+        }
+        fn provide(&mut self, key: String) -> Option<Value> {
+            let result = self.get_env_var(&key).cloned();
+            self.add_env_var(key, "".to_string().into_value(Span::unknown()));
+            result
+        }
+    }
+
+    impl EnvMaintainer for Stack {
+        fn maintain(&mut self, key: String, value: Value) {
+            self.add_env_var(key, value);
+        }
+        fn provide(&mut self, key: String) -> Option<Value> {
+            let engine_state = &EngineState::new();
+            let result = self.get_env_var(engine_state, &key).cloned();
+            self.remove_env_var(engine_state, &key);
+            result
+        }
+    }
+
+    /// For maintainer to notify PWD
+    /// When user changes current directory, maintainer calls set_pwd()
+    /// and PWD-per-drive call maintainer back to store it for the
+    /// env_var_for_drive.
+    /// TBD: If value can't be converted to String or the value is not valid for
+    /// windows path on a drive, should 'cd' or 'auto_cd' get warning message
+    /// that PWD-per-drive can't process the path value?
+    pub fn set_pwd<T: EnvMaintainer>(maintainer: &mut T, value: Value) -> Result<(), ShellError> {
+        match String::from_value(value.clone()) {
+            Ok(path_string) => {
+                if let Some(drive) = extract_drive_letter(&path_string) {
+                    maintainer.maintain(env_var_for_drive(drive), value.clone());
+                } else {
+                    // UNC Network share path (or any other format of path) must be mapped
+                    // to local drive, then CMD.exe can support current directory,
```

Comment:
why do we care about cmd.exe? i'm not following what you're saying here and why it is important.

Reply:
This means CMD.exe does not support cd \\server\share_path, so if PWD happend to be that format, just ignore is OK.

Reply:
Right, I realize cmd.exe doesn't support unc, but theoretically nushell could, like pwsh. This PR isn't about unc necessarily so I'm not saying it needs to be included.

Reply:
OK, will adjust the comment, currently just support standard C:, will extend the support range to other format if a drive letter can be recognized.

Reply:
Now when set_pwd() cannot detect drive letter, will return ShellError::InvalidValue